### ISSUE TITLE
Add on_click to Marker objects

### ISSUE THREAD

**Issue by rsignell-usgs** [jupyter-widgets/ipyleaflet#92]
I  tried out the new MarkerCluster capability in 0.5.3, and it works great in my notebook here:
https://gist.github.com/anonymous/d414d59e50e57e23436adba70fd48b35
but I'm struggling a bit with the interaction.  [BTW, this notebook should run anywhere, as it uses only web services to aquire data]

I was originally plotting markers using geoJSON, and the interaction I'm looking for is to return the name of  the marker to python when I click on it.   In cell [13], with geoJSON, I use `feature_layer.on_click` to specifying printing the name of the marker below the plot.  This works great:

![2017-12-19_14-51-54](https://user-images.githubusercontent.com/1872600/34175943-30ad689c-e4cc-11e7-9ae1-6a55fd5a174d.png)

I would like to do the same in cell [15], using `MarkerCluster`, but I cannot figure out how to get the name of the marker back on click:
![2017-12-19_14-53-27](https://user-images.githubusercontent.com/1872600/34176015-6c49e056-e4cc-11e7-9f84-915e8a7e0496.png)

There is no `feature_layer.on_click` for `MarkerCluster`, and I guess that make sense since I really just need interaction with the individual `Marker` objects, right?

But the `Marker` objects don't seem to have `on_click` either.  Would it be `on_trait_change`?

**Comment by jasongrout** [comment]
Should we add an on_click to Marker objects?

**Comment by rsignell-usgs** [comment]
Yes, unless `on_trait_change` or another existing method can be used to return marker properties on click.

**Comment by jasongrout** [comment]
I think the appropriate thing is to add an on_click handler, mirroring the button on_click handler in ipywidgets, or perhaps the on_click handler already in ipyleaflet.

What to give it a shot? It looks like the code is at https://github.com/ellisonbg/ipyleaflet/blob/0aa98732f3ca22fa6c442c5f1487b3aaba470710/ipyleaflet/leaflet.py#L427 and https://github.com/ellisonbg/ipyleaflet/blob/0aa98732f3ca22fa6c442c5f1487b3aaba470710/ipyleaflet/leaflet.py#L440 on the python side and the relevant js code is around https://github.com/ellisonbg/ipyleaflet/blob/0aa98732f3ca22fa6c442c5f1487b3aaba470710/js/src/jupyter-leaflet.js#L282

**Comment by jasongrout** [comment]
(here is the part in js that sends a message to the python side: https://github.com/ellisonbg/ipyleaflet/blob/0aa98732f3ca22fa6c442c5f1487b3aaba470710/js/src/jupyter-leaflet.js#L289)

**Comment by rsignell-usgs** [comment]
@jasongrout , ah, yes, an exercise for the student!   :smiley_cat: 
Okay, perhaps @ocefpaf and I can take a stab at it!

**Comment by ocefpaf** [comment]
> Okay, perhaps @ocefpaf and I can take a stab at it!

@rsignell-usgs I'll be offline for 2-3 days but after that I can give it a shot.

**Comment by SylvainCorlay** [comment]
@ocefpaf did you have the occasion to look at it?

We made some really good progress over the past couple of weeks, especially in cleaning up the general code base, and setting up the JupyterLab extension etc...

**Comment by ocefpaf** [comment]
Sorry @SylvainCorlay, I did not look at it. I don't think I'll have time for this anytime soon :-(

**Comment by martinRenou** [comment]
There is a `click` event in Lealfet https://leafletjs.com/reference-1.3.4.html#marker-click, so adding an `on_click` method to marker should be straightforward, the implementation will be like the `on_move` method.
And actually, because the `click` event is an event inherited from [Interactive Layer](https://leafletjs.com/reference-1.3.4.html#interactive-layer), we should be able to add this method directly to the `Layer` widget, and implementing all the mouse events should be straightforward too (dblclick, mousedown, mouseup, mouseover...).

**Comment by arkanoid87** [comment]
Is there any workaround to add this functionality now?
I'm trying with on_move as a workaround but seems not triggered, too

**Comment by martinRenou** [comment]
I can't think of a workaround... `on_move` is not triggered? Do you have any error message?

**Comment by arkanoid87** [comment]
Well, I was trying to turn my markers into GeoJSON to use the on_* handler there, but I'm not receiving any event here too. I think it is a problem on my side, the "CountriesGeoJSON.ipynb" demo works nicely

**Comment by martinRenou** [comment]
I wouldn't be surprised if GeoJSON events are not triggered properly... There are already some issues opened related to GeoJSON events, we need to give more love to this layer.
We'll work on the mouse events for the `Layer` widget this week, and I'll make a release of ipyleaflet on Friday.

**Comment by arkanoid87** [comment]
I've been doing some testing and the on_move event on Marker but seems not triggering even in a modified version of Primitives.ipnb

I've also tried to subscribe to on_msg but no messages there, too

**Comment by martinRenou** [comment]
I just tried it on the binder (last version of ipyleaflet) and `on_move` seems to work. Do you mind giving me an example so that I can replicate your issue?

**Comment by arkanoid87** [comment]
by adding *args, **kwargs to callback signature is working for me too!

I was not receiving any error and I've been trying with the callback signatures documented in the help(...)! It is misleading, but I'm happy to be able to continue now.

Thanks a lot

**Comment by martinRenou** [comment]
Cool :) Actually the callback should take parameters, you can print `args` and `kwargs` to see what is given. I guess it's a dictionary containing the new position of the `Marker` but I'm not entirely sure. So I you pass a callback that doesn't take parameters, it throws an exception, that doesn't get displayed in the notebook.

**Comment by arkanoid87** [comment]
It makes sense, but it was outside the common python behaviour (I'm new to jupyter). Now I'm working with functools.partial to keep a reference of the triggered Marker in the callback and it works, problem is that the callback seems before the update if the location on the leaflet side (I'm trying to push back the Marker on the original position as a workaround for on_click)

**Comment by martinRenou** [comment]
Annoying.. The callback is called before the location attribute has been updated, so your workaround won't work.. Do you mind waiting for a release? I'll add the mouse events like `on_click` tomorrow, and we'll make a release on Friday.

**Comment by arkanoid87** [comment]
Absolutely no problem, I was just trying to understand to (hopefully) contribute.
In the meantime I've re-tested the on_msg callback on Marker looking for any on_click feedback but seems not there, must be on the upper layer.

Thank you for taking care of this issue.

**Comment by martinRenou** [comment]
>  I was just trying to understand to (hopefully) contribute

That would be great! I created a github project today https://github.com/jupyter-widgets/ipyleaflet/projects/1. I added in the TODO list issues that seem to be easy enough for new contributors. If you start working on one of the issues I would be happy to help you.

Concerning the `on_msg` callback, you won't receive any `click` message, because only a `move` message is sent from the front-end https://github.com/jupyter-widgets/ipyleaflet/blob/master/js/src/jupyter-leaflet.js#L168.
